### PR TITLE
Support server-side pagination for Firestore

### PR DESCRIPTION
**Pull request by sharat87**

### CODE REVIEW

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java [hunk 1]
```diff
@@ -70,9 +72,22 @@ public FirestorePlugin(PluginWrapper wrapper) {
         private final Scheduler scheduler = Schedulers.elastic();
 
         @Override
+        @Deprecated
         public Mono<ActionExecutionResult> execute(Firestore connection,
                                                    DatasourceConfiguration datasourceConfiguration,
                                                    ActionConfiguration actionConfiguration) {
+            return executeParameterized(connection, null, datasourceConfiguration, actionConfiguration);
```
Comment (trishaanand) [inline app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:79]:
```suggestion
            return Mono.error(new AppsmithPluginException(AppsmithPluginError.PLUGIN_ERROR, "Unsupported Operation"));
```

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java [hunk 2]
```diff
@@ -240,16 +258,93 @@ public FirestorePlugin(PluginWrapper wrapper) {
...
+            if (property == null) {
+                return defaultValue;
+            }
+
+            final String value = property.getValue();
+            return value != null ? value : defaultValue;
+        }
+
+        private Mono<ActionExecutionResult> methodGetCollection(CollectionReference query, List<Property> properties, PaginationField paginationField) {
+            final String limitString = getPropertyAt(properties, 2, "10");
```
Comment (trishaanand) [inline app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:276]: Would it be possible to use static named fields for the property indexes to make the code more readable?

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java [hunk 3]
```diff
@@ -240,16 +258,93 @@ public FirestorePlugin(PluginWrapper wrapper) {
...
+                }
+            }
+
+            final Map<String, Object> startAfter = startAfterTemp;
+            final Map<String, Object> endBefore = endBeforeTemp;
+
+            if (paginationField != null && CollectionUtils.isEmpty(orderings)) {
+                return Mono.error(new AppsmithPluginException(
+                        AppsmithPluginError.PLUGIN_ERROR,
+                        "Cannot do pagination without specifying an ordering."
```
Comment (trishaanand) [inline app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:321]: Would this require external logging? Is it a plugin error or a user configuration issue?

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java [hunk 4]
```diff
@@ -240,16 +258,93 @@ public FirestorePlugin(PluginWrapper wrapper) {
...
+                ));
+            }
 
             return Mono.just(query)
                     // Apply ordering, if provided.
-                    .map(query1 -> StringUtils.isEmpty(orderBy) ? query1 : query1.orderBy(orderBy))
+                    // .map(query1 -> StringUtils.isEmpty(orderBy) ? query1 : query1.orderBy(orderBy))
```
Comment (trishaanand) [inline app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:327]: If possible, remove commented code.

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java (hunk 1 above)
Comment (sharat87) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:79]: Yeah, I was going to ask, can we just refactor the `executeParameterized` method into `execute` itself? Is there a reason we kept them both as separate methods?

Also, the tests call to `execute` today, which why I made the implementation this way. Just a shortcut :). If we should make this change, I have to change the tests as well. What say?

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java (hunk 3 above)
Comment (sharat87) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:321]: Yeah, user error. Changed error value.

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java (hunk 4 above)
Comment (sharat87) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java]: Done.

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java (hunk 1 above)
Comment (trishaanand) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:79]: Didnt want to change the plugins unnecessarily. executeParameterized was introduced to do special handling for variable substitution.

Comment (sharat87) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:79]: I would rather that we just refactor the `execute` method to have that extra argument. Unless there's two distinct uses for these two methods, it just looks like tech debt IMO. Either way, I'll make this change and update the tests and push.

Comment (trishaanand) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:79]: If we do this, we are adding the burden of doing variable substitution onto the plugin developer. Hence the variable substitution default code was added as part of plugin executor.
I agree with you that as code it actually would have been better. But I couldn't come up with a way where we don't pass on the preparation of the configurations to each plugin.

Comment (sharat87) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java:79]: Ah I see. So there is indeed a specific second purpose to the other method. My bad, that's the missing piece in my understanding. Thanks for the detail :)

File: app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java (hunk 2 above)
Comment (sharat87) [reply app/server/appsmith-plugins/firestorePlugin/src/main/java/com/external/plugins/FirestorePlugin.java]: Done and pushed.

**Review by trishaanand — APPROVED** [review]

**Pull request merged**